### DISCUSSION TITLE
Why is the molecule's value constructor called every time `useMolecule` is evaluated?

### DISCUSSION THREAD

**Discussion by mfurtak** [saasquatch/bunshi#40] [Q&A] [answered]
Hi folks,

I'm just getting started with Bunshi, and I'm a little confused by the behavior I'm seeing.

If I create a molecule with no molecule or scope dependencies:

```typescript
const expensiveObjMol = molecule(() => {
  // expensive object creation
  // ...
  return expensiveObj
})
```

And then I use that with:

```typescript
export const MyComponent = () => {
  const expensiveObj = useMolecule(expensiveObjMol);

  return (
   // ...
  );
};
```

My observation is that the molecule value constructor function is called every time the `useMolecule` is evaluated. This means that the work of creating a new expensive object is done every time `MyComponent` is rendered, _even though only the first value created is ever returned_ by `useMolecule`. So, if that's accurate, Bunshi is honoring the contract that I'll always get the same object reference, but it seems like it is also doing work that isn't desirable/useful.

My expectation would be that a dependency injection framework would avoid doing the work of creating new objects when they wouldn't be used.

Is my expectation wrong for Bunshi? Do I have the facts about the behavior right? What am I missing?

Thanks,
-Mike

**Comment by loganvolkers** [comment, marked as answer]
Hi @mfurtak I believe that this was reported as a bug in #39 and has been resolved in Bunshi 2.1.0

There are also new lifecycle hooks (`onMount` and `onUnmount`) to do more fine-grained control of memory-related operations.
